server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java

```diff
@@ -216,14 +216,37 @@ public void removeApplier(ClusterStateApplier applier) {
      * Add a listener for updated cluster states
      */
     public void addListener(ClusterStateListener listener) {
-        clusterStateListeners.add(listener);
+        clusterStateListeners.add(new ContextPreservingClusterStateListener(listener, threadPool.getThreadContext().newRestorableContext(true)));
+    }
+
+    private static class ContextPreservingClusterStateListener implements ClusterStateListener {
+        private final ClusterStateListener delegate;
+        private final Supplier<ThreadContext.StoredContext> supplier;
+
+        ContextPreservingClusterStateListener(ClusterStateListener delegate, Supplier<ThreadContext.StoredContext> supplier) {
+            this.delegate = delegate;
+            this.supplier = supplier;
+        }
+
+        @Override
+        public void clusterChanged(ClusterChangedEvent event) {
+            try (ThreadContext.StoredContext stored = supplier.get()) {
+                delegate.clusterChanged(event);
+            }
+        }
     }
 
     /**
      * Removes a listener for updated cluster states.
      */
-    public void removeListener(ClusterStateListener listener) {
-        clusterStateListeners.remove(listener);
+    public void removeListener(final ClusterStateListener listener) {
+        clusterStateListeners.removeIf(storedListener -> {
+            if (storedListener instanceof ContextPreservingClusterStateListener) {
```

Comment:
Could this ever be false now? I guess we could make it a `Collection<ContextPreservingClusterStateListener>`?

Reply:
Simplified